### PR TITLE
request handler added

### PR DESCRIPTION
**Pull request by stephanstapel**
I took the recommandation from #17 and did a straight-forward implementation.

I implemented it a simpler manner, not passing the request and response objects. Looking forward to your comments!

### CODE REVIEW

File: BricklinkSharp.Client/BricklinkClientFactory.cs [hunk 1]
```diff
@@ -30,19 +30,19 @@ namespace BricklinkSharp.Client;
 
 public static class BricklinkClientFactory
 {
-    private static IBricklinkClient Build(HttpClient httpClient, bool disposeHttpClient)
+    private static IBricklinkClient Build(HttpClient httpClient, bool disposeHttpClient, IBricklinkRequestHandler requestHandler = null)
```
Comment (gebirgslok) [inline BricklinkSharp.Client/BricklinkClientFactory.cs]:
```suggestion
    private static IBricklinkClient Build(HttpClient httpClient, bool disposeHttpClient, IBricklinkRequestHandler? requestHandler = null)
```

File: BricklinkSharp.Client/BricklinkClientFactory.cs [hunk 2]
```diff
@@ -30,19 +30,19 @@ namespace BricklinkSharp.Client;
...
     {
         var exchangeRatesService = new ExchangeRatesApiDotIo(httpClient);
-        return new BricklinkClient(httpClient, exchangeRatesService, disposeHttpClient);
+        return new BricklinkClient(httpClient, exchangeRatesService, disposeHttpClient, requestHandler);
     }
 
-    public static IBricklinkClient Build(HttpClient httpClient)
+    public static IBricklinkClient Build(HttpClient httpClient, IBricklinkRequestHandler requestHandler = null)
```
Comment (gebirgslok) [inline BricklinkSharp.Client/BricklinkClientFactory.cs]:
```suggestion
    public static IBricklinkClient Build(HttpClient httpClient, IBricklinkRequestHandler? requestHandler = null)
```

File: BricklinkSharp.Client/BricklinkClientFactory.cs [hunk 3]
```diff
@@ -30,19 +30,19 @@ namespace BricklinkSharp.Client;
...
     {
-        return Build(httpClient, false);
+        return Build(httpClient, false, requestHandler);
     }
 
-    public static IBricklinkClient Build()
+    public static IBricklinkClient Build(IBricklinkRequestHandler requestHandler = null)
```
Comment (gebirgslok) [inline BricklinkSharp.Client/BricklinkClientFactory.cs]:
```suggestion
    public static IBricklinkClient Build(IBricklinkRequestHandler? requestHandler = null)
```

File: BricklinkSharp.Client/BricklinkClient.cs [hunk 4]
```diff
@@ -51,19 +51,22 @@ internal sealed class BricklinkClient : IBricklinkClient
     private const string _originalBoxImageUrlTemplate = "//img.bricklink.com/ItemImage/ON/0/{0}.png";
 
     private readonly IExchangeRatesService _currencyRatesService;
+    private readonly IBricklinkRequestHandler _requestHandler;
```
Comment (gebirgslok) [inline BricklinkSharp.Client/BricklinkClient.cs]:
```suggestion
    private readonly IBricklinkRequestHandler? _requestHandler;
```

File: BricklinkSharp.Client/BricklinkClient.cs [hunk 5]
```diff
@@ -51,19 +51,22 @@ internal sealed class BricklinkClient : IBricklinkClient
...
     private readonly bool _disposeHttpClient;
 
     private bool _isDisposed;
 
-    public BricklinkClient(HttpClient httpClient,
+    public BricklinkClient(HttpClient httpClient,        
         IExchangeRatesService currencyRatesService,
-        bool disposeHttpClient)
+        bool disposeHttpClient,
+        IBricklinkRequestHandler requestHandler = null)
```
Comment (gebirgslok) [inline BricklinkSharp.Client/BricklinkClient.cs]:
```suggestion
        IBricklinkRequestHandler? requestHandler = null)
```

File: BricklinkSharp.Client/BricklinkClient.cs [hunk 6]
```diff
@@ -794,4 +838,14 @@ public void Dispose()
 
         return result;
     }
+
+
```
Comment (gebirgslok) [inline BricklinkSharp.Client/BricklinkClient.cs:842]: remove whitespace

File: BricklinkSharp.Client/BricklinkClient.cs [hunk 7]
```diff
@@ -794,4 +838,14 @@ public void Dispose()
...
+    private async void _measureRequest(CancellationToken cancellationToken = default)
```
Comment (gebirgslok) [inline BricklinkSharp.Client/BricklinkClient.cs]: Naming convention -> MeasureRequestAsync + async methods should return a task:
```suggestion
    private async Task MeasureRequestAsync(CancellationToken cancellationToken = default)
```

File: BricklinkSharp.Client/BricklinkClient.cs [hunk 8]
```diff
@@ -734,6 +776,7 @@ public void Dispose()
 
         var url = new Uri(_baseUri, $"coupons/{couponId}").ToString();
 
+        _measureRequest(cancellationToken);
```
Comment (gebirgslok) [inline BricklinkSharp.Client/BricklinkClient.cs]: MeasureRequest should be awaited?
```suggestion
        await MeasureRequestAsync(cancellationToken);
```

File: BricklinkSharp.Client/BricklinkClient.cs [hunk 9]
```diff
@@ -794,4 +838,14 @@ public void Dispose()
 
         return result;
     }
+
+
+    private async void _measureRequest(CancellationToken cancellationToken = default)
+    {
+        if (this._requestHandler != null)
+        {
+            CancellationTokenSource source = new CancellationTokenSource();
```
Comment (gebirgslok) [inline BricklinkSharp.Client/BricklinkClient.cs]: source is unused?

**Review by gebirgslok — CHANGES_REQUESTED** [review]

File: BricklinkSharp.Client/BricklinkClient.cs (hunk 9 above)
Comment (stephanstapel) [reply BricklinkSharp.Client/BricklinkClient.cs]: removed in commit [9c62133](https://github.com/gebirgslok/BricklinkSharp/pull/19/commits/9c62133c17d90d5bd1db2416d407e86a6c9aecb6)

**Comment by stephanstapel** [comment]
The code needs a couple change to make the async hook work.
It looks like not all functions that return Task and are named *Async are in fact async in BricklinkClient. Is there a reason for this?
Here is an example:

grafik

**Comment by gebirgslok** [comment]
> The code needs a couple change to make the async hook work. It looks like not all functions that return Task and are named *Async are in fact async in BricklinkClient. Is there a reason for this? Here is an example:
> 
> grafik

You can make them async (but don't forget to await the HttpClient call) :)

File: BricklinkSharp.Client/BricklinkClient.cs [hunk 10]
```diff
@@ -840,12 +840,11 @@ public void Dispose()
     }
 
 
-    private async void _measureRequest(CancellationToken cancellationToken = default)
+    private async Task MeasureRequestAsync(CancellationToken cancellationToken = default)
     {
-        if (this._requesHandler != null)
+        if (this._requestHandler != null)
```
Comment (gebirgslok) [inline BricklinkSharp.Client/BricklinkClient.cs]: One last issue (sorry), but the underscore for private fields is used to not require "this" ;)
```suggestion
        if (_requestHandler != null)
```

**Review by gebirgslok — APPROVED** [review]

**Comment by gebirgslok** [comment]
Thanks for your contribution! I'll release a 1.6.0 version tomorrow 👍

**Pull request merged**